src/NHibernate/Persister/Entity/SubqueryPropertyMapping.cs

```diff
@@ -0,0 +1,346 @@
+			{
+				return columns;
+			}
+
+			throw new QueryException(string.Format("could not resolve property: {0} of: {1}", propertyName, "SubQuery"));
+		}
+
+		public List<string> GetPropertiesColumns(string alias)
+		{
+			var columns = new List<string>();
+			var processedElements = new HashSet<FromElement>();
+			foreach (var expression in _selectClause.NonScalarExpressions)
+			{
+				var fromElement = expression.FromElement;
+				if (fromElement == null || fromElement.IsFetch || !processedElements.Add(fromElement))
+				{
+					continue;
+				}
+
+				var fragment = fromElement.GetPropertiesSelectFragment(fromElement.EntitySuffix, alias);
+				if (fromElement is JoinSubqueryFromElement)
+				{
+					columns.AddRange(fragment.GetColumnAliases());
+				}
+				else
+				{
+					columns.AddRange(fragment.GetColumnAliases().Select(o => StringHelper.Qualify(alias, o)));
+				}
+			}
+
+			if (_selectClause.ColumnNames != null)
+			{
+				foreach (var scalarColumns in _selectClause.ColumnNames)
+				{
+					columns.AddRange(QualifyColumns(scalarColumns, alias));
+				}
+			}
+
+			return columns;
+		}
+
+		public List<string> GetIdentifiersColumns(string alias)
+		{
+			var columns = new List<string>();
+			var processedElements = new HashSet<FromElement>();
+			foreach (var expression in _selectClause.NonScalarExpressions)
+			{
+				var fromElement = expression.FromElement;
+				if (fromElement == null ||
+				    fromElement.FromClause.IsScalarSubQuery ||
+				    fromElement.IsFetch ||
+				    !processedElements.Add(fromElement))
+				{
+					continue;
+				}
+
+				var fragment = fromElement.GetIdentifierSelectFragment(fromElement.EntitySuffix, alias);
+				if (fragment == null)
+				{
+					continue; // Subquery with scalar select
+				}
+
+				if (fromElement is JoinSubqueryFromElement)
+				{
+					columns.AddRange(fragment.GetColumnAliases());
+				}
+				else
+				{
+					columns.AddRange(fragment.GetColumnAliases().Select(o => StringHelper.Qualify(alias, o)));
+				}
+			}
+
+			return columns;
+		}
+
+		private bool TryGetColumns(string alias, string propertyName, out string[] columns)
+		{
+			var paths = propertyName.Split(StringHelper.Dot);
+			var propertyAlias = paths[0];
+			if (paths.Length == 1 && _propertyColumns.TryGetValue(propertyAlias, out columns))
+			{
+				columns = QualifyColumns(columns, alias);
+				return true;
+			}
+
+			if (_propertyAliasMappings.TryGetValue(propertyAlias, out var mapping))
+			{
+				var suffix = _propertyMappingSuffixes[propertyAlias];
+				if (
+					(paths.Length > 1 && TryGetQualifiedColumns(mapping, JoinPaths(paths.Skip(1)), alias, suffix, out columns)) ||
+					(paths.Length == 1 && TryGetQualifiedColumns(mapping, EntityPersister.EntityID, alias, suffix, out columns))
+				)
+				{
+					return true;
+				}
+			}
+
+			if (propertyName == EntityPersister.EntityID)
+			{
+				columns = GetIdentifiersColumns(alias).ToArray();
+				return columns.Length > 0;
```

Comment:
Should we cache results in `_propertyColumns`?

Reply:
`_propertyColumns` contains only columns names, where `GetIdentifiersColumns` returns qualified columns. As alias may be different (nested subqueries) I rather cached non scalar `FromElement`s to avoid using a `HashSet` in `GetPropertiesColumns` and `GetIdentifiersColumns` methods.